lib/testing/matrix.js

```diff
@@ -0,0 +1,209 @@
+'use strict'; // eslint-disable-line strict
+
+/**
+* Running tests with multiple configurations using varying parameters
+* @constructor
+*/
+class TestMatrix {
+    /**
+    * Builds a new testing Matrix
+    *
+    * @param {Object.<string, string|string[]>} params - Argument of the matrix
+    *
+    * @returns {TestMatrix} new testing Matrix
+    */
+    constructor(params) {
+        /**
+        * Create deep copy for prevent the user to modify the object outside
+        * the class
+        **/
+        this.params = Object.assign({}, params);
+        this.paramsToString = JSON.stringify(this.params);
+        this.callback = null;
+        this.elementsToSpecialize = [];
+        this.description = '';
+        this.listOfExceptions = [];
+    }
+
+    /**
+    * Callback used for prevent mocha that the test is over
+    * Must be call in the end of every test
+    * @callback Done
+    *
+    * @return {undefined}
+    */
+    /**
+    * Callback used for a generated matrix
+    * Used by @generate and @if method
+    *
+    * @callback GenerateMatrixCallback
+    * @param {TestMatrix} matrix - The following generate matrix.
+    * @param {Done} done - Callback for prevent mocha that we finish
+    * the current test. This callback will be empty if not all elements are
+    * specialize
+    *
+    * @return {undefined}
+    */
+
+    /**
+    * Generate testing matrix, call execute method to run callback
+    *
+    * @param {string[]} elementsToSpecialize - key element we want to
+    *                    specialize on the matrix
+    * @param {GenerateMatrixCallback} callback - callback function
+    *                    who want to be call on execute method
+    *                    for the current generate matrix
+    * @param {string} description - description of the test
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    generate(elementsToSpecialize, callback, description) {
+        this.elementsToSpecialize = elementsToSpecialize;
+        this.callback = callback;
+        this.description = typeof description === 'undefined'
+        ? ''
+        : description;
+        return this;
+    }
+
+    /**
+    * Method to prevent an exception during the tests.
+    *
+    * @param {Object.<string, string[]>} exception - The keys of the hash table
+    is the element we want to do an exception, the value
```

Comment:
An exception is a special case when we have specific items on the matrix.
Should we rename at "Special case" instead of exception ?